### PR TITLE
[TEP-0076] Add indexing into array for pipeline params reference

### PR DESCRIPTION
**Pull request by Yongxuanzhang**
# Changes
This commit provides the indexing into array for params for pipeline level reference and gated by
alpha feature flag. Before this commit we can only refer to the whole
array param, with this feature we can refer to array's element such as
$(params.param-name[i]).

/kind feature

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Release notes block below has been filled in
(if there are no user facing changes, use release note "NONE")

# Release Notes

```release-note
Indexing into array for pipeline params is now an alpha feature, element of array params can be accessed via $(params.param-name[i]).
```

### CODE REVIEW

**Comment by Yongxuanzhang** [comment]
/assign @ywluogg 
Realized that this is independent but also part of [tep0076](https://github.com/tektoncd/community/blob/main/teps/0076-array-result-types.md#indexing-into-arrays-with-variable-replacement)

File: docs/variables.md
Comment (ywluogg) [inline docs/variables.md:46]: I think other rows are in separate rows.

File: pkg/reconciler/pipelinerun/resources/apply.go
Comment (ywluogg) [inline pkg/reconciler/pipelinerun/resources/apply.go:60]: nit: pattern + "[%d]"

Comment (Yongxuanzhang) [reply pkg/reconciler/pipelinerun/resources/apply.go]: Thanks! This is something I'm not aware of before

File: docs/variables.md
Comment (ywluogg) [reply docs/variables.md:46]: Oh! Glad that you verified it. I was seeing on Github preview. You can close this thread

File: pkg/reconciler/pipelinerun/resources/apply.go
Comment (ywluogg) [reply pkg/reconciler/pipelinerun/resources/apply.go:60]: That's weird - can you show me the errors?

File: examples/v1beta1/taskruns/alpha/param_array_indexing.yaml
Comment (ywluogg) [inline examples/v1beta1/taskruns/alpha/param_array_indexing.yaml:22]: Can you add another step in this to test `args: ["echo $(params.array-to-echo[0])"]`? or `args: ["echo $(params.array-to-echo[0])"] > $(workspaces.testworkspace.path)/testfile.txt` something like that?

**Review by ywluogg — APPROVED** [review]
Except for the current open comments, other LGTM

File: pkg/reconciler/pipelinerun/resources/apply.go
Comment (Yongxuanzhang) [reply pkg/reconciler/pipelinerun/resources/apply.go:60]: gofmt will remove these spaces. So if I add these it will fail the lint check
I think it is in this log.
https://storage.googleapis.com/tekton-prow/pr-logs/pull/tektoncd_pipeline/4855/pull-tekton-pipeline-build-tests/1524386665529872384/build-log.txt

Comment (ywluogg) [reply pkg/reconciler/pipelinerun/resources/apply.go:60]: Oh! I guess that's the new formatting then. You can close this thread now

Comment (Yongxuanzhang) [reply pkg/reconciler/pipelinerun/resources/apply.go:60]: Got it, thanks!

File: examples/v1beta1/taskruns/alpha/param_array_indexing.yaml
Comment (Yongxuanzhang) [reply examples/v1beta1/taskruns/alpha/param_array_indexing.yaml:22]: this seems not the correct syntax, did you mean
`args: ["echo $(params.array-to-echo[0])"]`?

Comment (ywluogg) [reply examples/v1beta1/taskruns/alpha/param_array_indexing.yaml:22]: Ah yes. I saw you added that. You can close this thread now

**Comment by ywluogg** [comment]
Link this to #4723

**Comment by ywluogg** [comment]
Hi Yongxuan, do you need to find someone to give LGTM for this PR? I see this hasn't been revisited for a long time

**Comment by Yongxuanzhang** [comment]
> Hi Yongxuan, do you need to find someone to give LGTM for this PR? I see this hasn't been revisited for a long time

Oh thanks for reminding me of this one. I will resolve those conflicts!

**Comment by Yongxuanzhang** [comment]
https://github.com/tektoncd/pipeline/pull/4857 is blocking this PR's work since it will fail the validation webhook

File: pkg/reconciler/pipelinerun/resources/apply.go
Comment (dibyom) [inline pkg/reconciler/pipelinerun/resources/apply.go:60]:
```suggestion
					// TODO(#4723): Validate the array reference is not out of bound
```

Comment (dibyom) [reply pkg/reconciler/pipelinerun/resources/apply.go:60]: What happens today if the array is out of bound?

File: pkg/reconciler/taskrun/resources/apply_test.go
Comment (dibyom) [reply pkg/reconciler/taskrun/resources/apply_test.go:347]: how hard would it be to add this validation? I think silently failing without errors can be confusing for users

File: pkg/reconciler/taskrun/resources/apply_test.go
Comment (dibyom) [inline pkg/reconciler/taskrun/resources/apply_test.go:984]: Do we really need all of these mutations to test indexing?

File: pkg/reconciler/pipelinerun/resources/apply.go
Comment (Yongxuanzhang) [reply pkg/reconciler/pipelinerun/resources/apply.go:60]: I will remove this todo

File: pkg/reconciler/taskrun/resources/apply_test.go
Comment (Yongxuanzhang) [reply pkg/reconciler/taskrun/resources/apply_test.go:347]: Can we keep the validation in another PR? The release deadline is close

Comment (lbernick) [reply pkg/reconciler/taskrun/resources/apply_test.go:347]: I agree with Dibyo, I wouldn't want to hold off on adding validation just to get the feature out the door. Can you explain more about what the challenge is here?

File: pkg/substitution/substitution.go
Comment (lbernick) [inline pkg/substitution/substitution.go:36]: could you please use a helper function rather than exporting this?

File: docs/variables.md
Comment (Yongxuanzhang) [reply docs/variables.md:47]: That's weird, I remember I removed this

File: pkg/reconciler/pipelinerun/resources/validate_params.go
Comment (Yongxuanzhang) [reply pkg/reconciler/pipelinerun/resources/validate_params.go:108]: yeah ok, I will try to extract a func from this

File: pkg/reconciler/pipelinerun/resources/validate_params.go
Comment (Yongxuanzhang) [reply pkg/reconciler/pipelinerun/resources/validate_params.go]: sure, I moved them to substitution and warp with funcs

Comment (Yongxuanzhang) [reply pkg/reconciler/pipelinerun/resources/validate_params.go]: oh I see, I removed the pipeline ones, just fixed

File: pkg/reconciler/pipelinerun/resources/validate_params.go
Comment (lbernick) [inline pkg/reconciler/pipelinerun/resources/validate_params.go:189]: can this string parsing be moved into the substitution package or elsewhere?

File: pkg/reconciler/pipelinerun/resources/validate_params_test.go
Comment (Yongxuanzhang) [inline pkg/reconciler/pipelinerun/resources/validate_params_test.go:484]: @lbernick is this the test case?

File: pkg/substitution/substitution.go
Comment (lbernick) [inline pkg/substitution/substitution.go:329]: typo: leftmost

**Review by lbernick — APPROVED** [review]

**Comment by dibyom** [comment]
Thanks @Yongxuanzhang 
/lgtm

**Pull request merged**
